File: tests/Emphasize.UnitTests/FormatDefinitionTests.vb
Imports Microsoft.VisualStudio.Text.Classification
Imports Microsoft.VisualStudio.Utilities
Imports System.Reflection


Public Class FormatDefinitionTests

    <Fact()>
    Public Sub AllDefinitionsHaveUniqueNames()
        Dim names As New HashSet(Of String)(StringComparer.OrdinalIgnoreCase)


        For Each type In GetFormatDefinitionTypes().Select(Function(x) x.Cast(Of Type).First())
            names.Add(GetName(type))
        Next type

        Assert.Equal(7, names.Count)
    End Sub


    <MemberData(NameOf(GetFormatDefinitionTypes))>
    <Theory()>
    Public Sub AllDefinitionNamesStartWithComment(type As Type)
        Assert.StartsWith("Comment - ", GetName(type))
    End Sub


    <MemberData(NameOf(GetFormatDefinitionTypes))>
    <Theory()>
    Public Sub AllDefinitionsExportUsingTheirOwnName(type As Type)
        Dim name As String


        name = GetName(type)

        Assert.Equal(name, type.GetCustomAttribute(Of NameAttribute)().Name)
        Assert.Equal(name, type.GetCustomAttribute(Of ClassificationTypeAttribute)().ClassificationTypeNames)
    End Sub


    <MemberData(NameOf(GetFormatDefinitionTypes))>
    <Theory()>
    Public Sub AllDefinitionsExportUseTheirNameAsTheirDisplayName(type As Type)
        Dim name As String
        Dim definition As ClassificationFormatDefinition


        name = GetName(type)
        definition = DirectCast(Activator.CreateInstance(type), ClassificationFormatDefinition)

        Assert.Equal(name, definition.DisplayName)
    End Sub


    <MemberData(NameOf(GetFormatDefinitionTypes))>
    <Theory()>
    Public Sub AllDefinitionsDoWhatTheySayTheyDo(type As Type)
        Dim definition As ClassificationFormatDefinition
        Dim name As String


        definition = DirectCast(Activator.CreateInstance(type), ClassificationFormatDefinition)

        name = GetName(type)

        If name.IndexOf("bold", StringComparison.OrdinalIgnoreCase) >= 0 Then
            Assert.True(definition.IsBold)
        Else
            Assert.Null(definition.IsBold)
        End If

        If name.IndexOf("italic", StringComparison.OrdinalIgnoreCase) >= 0 Then
            Assert.True(definition.IsItalic)
        Else
            Assert.Null(definition.IsItalic)
        End If

        If name.IndexOf("code", StringComparison.OrdinalIgnoreCase) >= 0 Then
            Assert.NotNull(definition.BackgroundBrush)
            Assert.NotNull(definition.BackgroundOpacity)
        Else
            Assert.Null(definition.BackgroundBrush)
            Assert.Null(definition.BackgroundOpacity)
        End If
    End Sub


    Public Shared Iterator Function GetFormatDefinitionTypes() As IEnumerable(Of Object())
        Yield {GetType(FormatDefinitions.Bold)}
        Yield {GetType(FormatDefinitions.BoldItalic)}
        Yield {GetType(FormatDefinitions.BoldCode)}
        Yield {GetType(FormatDefinitions.BoldItalicCode)}
        Yield {GetType(FormatDefinitions.Italic)}
        Yield {GetType(FormatDefinitions.ItalicCode)}
        Yield {GetType(FormatDefinitions.Code)}
    End Function


    Private Shared Function GetName(definitionType As Type) As String
        Dim flags As BindingFlags


        flags = BindingFlags.Public Or BindingFlags.Static

        Return DirectCast(definitionType.GetField("Name", flags).GetValue(Nothing), String)
    End Function

End Class

File: tests/Emphasize.UnitTests/FormatDefinitionTests.cs
using Microsoft.VisualStudio.Text.Classification;
using Microsoft.VisualStudio.Utilities;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Xunit;


namespace Emphasize {

    public class FormatDefinitionTests {

        [Fact]
        public void AllDefinitionsHaveUniqueNames() {
            HashSet<string> names = new(StringComparer.OrdinalIgnoreCase);


            foreach (var type in GetFormatDefinitionTypes().Select((x) => x.Cast<Type>().First())) {
                names.Add(GetName(type));
            }

            Assert.Equal(7, names.Count);
        }


        [Theory]
        [MemberData(nameof(GetFormatDefinitionTypes))]
        public void AllDefinitionNamesStartWithComment(Type type) {
            Assert.StartsWith("Comment - ", GetName(type));
        }


        [Theory]
        [MemberData(nameof(GetFormatDefinitionTypes))]
        public void AllDefinitionsExportUsingTheirOwnName(Type type) {
            string name;


            name = GetName(type);

            Assert.Equal(name, type.GetCustomAttribute<NameAttribute>().Name);
            Assert.Equal(name, type.GetCustomAttribute<ClassificationTypeAttribute>().ClassificationTypeNames);
        }


        [Theory]
        [MemberData(nameof(GetFormatDefinitionTypes))]
        public void AllDefinitionsExportUseTheirNameAsTheirDisplayName(Type type) {
            string name;
            ClassificationFormatDefinition definition;


            name = GetName(type);
            definition = (ClassificationFormatDefinition)Activator.CreateInstance(type);

            Assert.Equal(name, definition.DisplayName);
        }


        [Theory]
        [MemberData(nameof(GetFormatDefinitionTypes))]
        public void AllDefinitionsDoWhatTheySayTheyDo(Type type) {
            ClassificationFormatDefinition definition;
            string name;


            definition = (ClassificationFormatDefinition)Activator.CreateInstance(type);
            name = GetName(type);

            if (name.IndexOf("bold", StringComparison.OrdinalIgnoreCase) >= 0) {
                Assert.True(definition.IsBold);
            } else {
                Assert.Null(definition.IsBold);
            }

            if (name.IndexOf("italic", StringComparison.OrdinalIgnoreCase) >= 0) {
                Assert.True(definition.IsItalic);
            } else {
                Assert.Null(definition.IsItalic);
            }

            if (name.IndexOf("code", StringComparison.OrdinalIgnoreCase) >= 0) {
                Assert.NotNull(definition.BackgroundBrush);
                Assert.NotNull(definition.BackgroundOpacity);
            } else {
                Assert.Null(definition.BackgroundBrush);
                Assert.Null(definition.BackgroundOpacity);
            }
        }


        public static IEnumerable<object[]> GetFormatDefinitionTypes() {
            yield return new object[] { typeof(FormatDefinitions.Bold) };
            yield return new object[] { typeof(FormatDefinitions.BoldItalic) };
            yield return new object[] { typeof(FormatDefinitions.BoldCode) };
            yield return new object[] { typeof(FormatDefinitions.BoldItalicCode) };
            yield return new object[] { typeof(FormatDefinitions.Italic) };
            yield return new object[] { typeof(FormatDefinitions.ItalicCode) };
            yield return new object[] { typeof(FormatDefinitions.Code) };
        }


        private static string GetName(Type definitionType) {
            return definitionType.GetCustomAttribute<NameAttribute>().Name;
        }

    }

}